### DISCUSSION TITLE
Can't use Livewire for package.

### DISCUSSION THREAD

**Discussion by ArtMin96** [livewire/livewire#4494] [1. Help]
I need to create a livewire component for my package. How can I create this with a command line?

Creating a package is very difficult for me when I first try to create a Livewire package. When I register the package ```Livewire\Livewire::component('componentName', ComponentName::class)``` I get the following error **Class 'Livewire\Livewire' not found**

I do not know how to create a component for a package with a command. Is there a tutorial to get started with at least a minimal setup?

**Comment by nuernbergerA** [comment]
Hey @ArtMin96,

in order to use livewire in your package, you need to require it `composer require livewire/livewire`

https://laravel-livewire.com/docs/2.x/package-dev

**Comment by ArtMin96** [comment]
I did it. I tried ```composer dump-autoload```, ```php artisan livewire: discover``` and no changes.

I tried to call the components differently, but in that case I get the following error.

```php
$this->registerLivewireComponentDirectory(__DIR__ . '/Http/Livewire', 'Vendor\\PackageName\\Http\\Livewire', 'package.core');

protected function registerLivewireComponentDirectory($directory, $namespace, $aliasPrefix = '')
{
    $filesystem = new Filesystem();

    if (! $filesystem->isDirectory($directory)) {
        return;
    }

    collect($filesystem->allFiles($directory))
        ->map(function (SplFileInfo $file) use ($namespace) {
            return (string) Str::of($namespace)
        ->append('\\', $file->getRelativePathname())
        ->replace(['/', '.php'], ['\\', '']);
    })
    ->filter(function ($class) {
        return is_subclass_of($class, Component::class) && ! (new ReflectionClass($class))->isAbstract();
    })
    ->each(function ($class) use ($namespace, $aliasPrefix) {
        $alias = Str::of($class)
                    ->after($namespace . '\\')
                    ->replace(['/', '\\'], '.')
                    ->prepend($aliasPrefix)
                    ->explode('.')
                    ->map([Str::class, 'kebab'])
                    ->implode('.');

                Livewire::component($alias, $class);
    });
}
```

Error: **Class 'Livewire\Component' not found**.

nor can I create a component command.

**Comment by joshhanley** [comment]
@ArtMin96 where are you trying to register the components? They should be in a service provider. If you are using a service provider, can you share the all of code for it?

I have a demo package setup here if you are interested in looking at it https://github.com/joshhanley/livewire-package-demo.

Hope this helps!

**Comment by ArtMin96** [comment]
Thanks for your help guys

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by ArtMin96** [comment]
Hello. This is my second time trying to use Livewire in my package and the same problem has occurred again.
Why does the package not see livewire? Even @livewireStyles and @livewireScripts appear as text in the browser.

Livewire required.

**Class 'Livewire\Livewire' not found**

```php

namespace Minasyans\LaravelInstaller;

use Illuminate\Support\ServiceProvider;
use Livewire\Livewire;
use Minasyans\LaravelInstaller\Commands\StarterKitCommand;
use Minasyans\LaravelInstaller\Middleware\CanConfigureApplication;
use Minasyans\LaravelInstaller\Middleware\CanInstall;
use Minasyans\LaravelInstaller\Middleware\CanUpdate;
use Minasyans\LaravelInstaller\Middleware\CheckPermissions;
use Minasyans\LaravelInstaller\Middleware\CheckRequirements;
use Minasyans\LaravelInstaller\Http\Livewire\TestComponent;

class LaravelInstallerServiceProvider extends ServiceProvider
{
    /**
     * Indicates if loading of the provider is deferred.
     *
     * @var bool
     */
    protected $defer = false;

    /**
     * Perform post-registration booting of services.
     *
     * @return void
     */
    public function boot(): void
    {
         $this->loadTranslationsFrom(__DIR__.'/../resources/lang', 'laravel-installer');
         $this->loadViewsFrom(__DIR__.'/../resources/views', 'laravel-installer');
        // $this->loadMigrationsFrom(__DIR__.'/../database/migrations');
         $this->loadRoutesFrom(__DIR__.'/routes.php');

        $this->app['router']->aliasMiddleware('install', CanInstall::class);
        $this->app['router']->aliasMiddleware('update', CanUpdate::class);
        $this->app['router']->aliasMiddleware('checkRequirements', CheckRequirements::class);
        $this->app['router']->aliasMiddleware('checkPermissions', CheckPermissions::class);
        $this->app['router']->aliasMiddleware('configureApplication', CanConfigureApplication::class);

        // Publishing is only necessary when using the CLI.
        if ($this->app->runningInConsole()) {
            $this->bootForConsole();
        }

        $this->commands([
            StarterKitCommand::class
        ]);

        Livewire::component('test-component', TestComponent::class);
    }

    ...
```

And component 

```php

namespace Minasyans\LaravelInstaller\Http\Livewire;

use Livewire\Component;

class TestComponent extends Component
{
    public function render()
    {
        return view('test-component');
    }
}
```

**Comment by ChaerilM** [comment]
@ArtMin96 maybe composer update, or check https://forum.laravel-livewire.com/t/resolved-livewirescripts-and-livewirestyles-showing-on-page/147/12

**Comment by ArtMin96** [comment]
@ChaerilM Thanks for the reply but still not working.
I required livewire only in the package, can this be a problem?
Here is a screenshot of the project structure

![image](https://user-images.githubusercontent.com/29732308/148647689-265e0ed5-06c7-4930-b6af-3c70a8490ac8.png)

### Layout code

![Klj56XM](https://user-images.githubusercontent.com/29732308/148647876-ad38cb34-40c1-4a70-9c7e-25eb4652887f.png)

### And webpage screenshot

![screencapture-127-0-0-1-8000-install-2022-01-08-18_25_05](https://user-images.githubusercontent.com/29732308/148647816-03a43d63-306d-4e3e-9e29-252a5137463e.png)

**Comment by ArtMin96** [comment]
I think the problem is that I can not use livewire only in the package. When I required livewire in the project `@livewireStyles` and `@livewireScripts` disappeared which means it worked. But when I call a component it looks up the app/Http/Livewire/ directory and gives an error. Message **The "/var/www/project_dir/app/Http/Livewire/" directory does not exist**. How can I use livewire if it is required only in the package?

**Comment by ArtMin96** [comment]
Now I know the problem is not with livewire. I just need help to get out of this situation.

**Comment by ArtMin96** [comment]
All fixed.

### PackageServiceProvider

```php
/**
* Register any package services.
*
* @return void
*/
public function register(): void
{
    ...

    $this->registerLivewireComponentDirectory(__DIR__ . '/Http/Livewire', 'Minasyans\\LaravelInstaller\\Http\\Livewire', '');
}

protected function registerLivewireComponentDirectory($directory, $namespace, string $aliasPrefix = '')
{
    $filesystem = new Filesystem();

    if (! $filesystem->isDirectory($directory)) {
        return;
    }

    collect($filesystem->allFiles($directory))
        ->map(function (SplFileInfo $file) use ($namespace) {
                return (string) Str::of($namespace)
                    ->append('\\', $file->getRelativePathname())
                    ->replace(['/', '.php'], ['\\', '']);
            })
            ->filter(function ($class) {
                return is_subclass_of($class, Component::class) && ! (new ReflectionClass($class))->isAbstract();
            })
            ->each(function ($class) use ($namespace, $aliasPrefix) {
                $alias = Str::of($class)
                    ->after($namespace . '\\')
                    ->replace(['/', '\\'], '.')
                    ->prepend($aliasPrefix)
                    ->explode('.')
                    ->map([Str::class, 'kebab'])
                    ->implode('.');

                Livewire::component($alias, $class);
            });
}
```

### Component

```php
class TestComponent extends Component
{
    public function render()
    {
        return view('laravel-installer::livewire.test-component');
    }
}
```

**Comment by joshhanley** [comment]
This issue has been converted to a discussion as we are now using discussions to track bugs and ideas.
See this announcement for more details https://github.com/livewire/livewire/discussions/3147

**Comment by metasanjaya** [comment]
Hello, I tried to use [this solution](https://github.com/livewire/livewire/discussions/4494#discussioncomment-1936646) but error occured. I use Livewire 3

```
In Container.php line 906:

  Target class [livewire] does not exist.


In Container.php line 904:

  Class "livewire" does not exist
```

<img width="788" alt="Screenshot 2024-04-04 at 21 12 53" src="https://github.com/livewire/livewire/assets/239236/5edf112d-5599-4c5f-bf92-8237219fdbd6">


Please help.
Thank you.

**Reply by ArtMin96** [reply]
Check the class name.
Target class **livewire** does not exist.
Where is the `livewire` class that starts with a lowercase **l**?

**Reply by metasanjaya** [reply]
I am sorry, is it because my package is called before livewire package?

on my package service provider on register method I add this
`Livewire::component('mypackage-component', MyPackage::class);`
and trigger the error. But if I remove or comment it, the error gone.

how to make my package called after livewire package ?

Thank you
